### PR TITLE
Jupyter canvas

### LINKED ISSUE #176: Jupyter widget based on jupyter_rfb

### PR DESCRIPTION
**Pull request by almarklein**
Closes #176

I'm going to leave this open until I've updated pygfx, so I can make sure that it works well for pygfx too.

jupyter_rfb3

### CODE REVIEW

File: wgpu/gui/jupyter.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return int(self._logical_size[0] * self._pixel_ratio), int(
+            self._logical_size[1] * self._pixel_ratio
+        )
+
+    def set_logical_size(self, width, height):
+        self.css_width = f"{width}px"
+        self.css_height = f"{height}px"
+
+    def close(self):
+        RemoteFrameBuffer.close()
```
Comment (Korijn) [inline wgpu/gui/jupyter.py:56]: Don't you need to pass `self` here?

**Comment by Korijn** [comment]
Can you tell me a little bit about how this works if someone has multiple cells with rendered scenes in each cell?

File: wgpu/gui/jupyter.py (hunk 1 above)
Comment (almarklein) [reply wgpu/gui/jupyter.py:56]: oooh nice catch!

**Comment by almarklein** [comment]
> Can you tell me a little bit about how this works if someone has multiple cells with rendered scenes in each cell?

Good question. Each cell would probably create a new canvas with a new renderer. So the device would be shared (we should probably make that so), but each canvas comes with its own pipeline(s). I should make a notebook with multiple cells to get a better feel. Will do when we can use pygfx.

**Comment by almarklein** [comment]
image

**Comment by almarklein** [comment]
BTW: you may see in the screenshots that the colors are a bit off. This has been fixed too by using the appopriate texture format.

**Comment by Korijn** [comment]
What happens if you leave out the animation loop and just render once? Does that work?

**Comment by almarklein** [comment]
Yes. Then a redraw is only requested when the canvas is resized (in the browser) like with normal windows.

**Comment by almarklein** [comment]
Examples:  https://jupyter-rfb.readthedocs.io/en/latest/examples/

**Pull request merged**